packages/synthetix-main/contracts/submodules/account/AccountBase.sol

```diff
@@ -0,0 +1,112 @@
+    uint16 private constant _PERMISSION_MANAGE_LOCKING = uint16(1) << 6;
+
+    // ---------------------------------------
+    // Enums
+    // ---------------------------------------
+
+    // ---------------------------------------
+    // Errors
+    // ---------------------------------------
+
+    error NotAuthorized(uint256 account, address addr, uint16 permission);
+
+    // ---------------------------------------
+    // Events
+    // ---------------------------------------
+
+    event PermissionGranted(uint256 account, address authorized, uint16 permission);
+    event PermissionRevoked(uint256 account, address authorized, uint16 permission);
+    event PermissionsSet(uint256 account, address authorized, uint16 packedPermissions);
+    event PermissionsCleared(uint256 account, address authorized);
+
+    // ---------------------------------------
+    // Helpers
+    // ---------------------------------------
+
+    function _isInitialized() internal view override returns (bool) {
+        return _accountStore().initialized;
+    }
+
+    function _hasPermission(
+        uint256 account,
+        address authorized,
+        uint16 permission
+    ) internal view returns (bool) {
+        return _accountStore().accountDelegations[account][authorized] & permission != 0;
+    }
+
+    function _grantPermission(
+        uint256 account,
+        address authorized,
+        uint16 permission
+    ) internal {
+        _accountStore().accountDelegations[account][authorized] =
+            _accountStore().accountDelegations[account][authorized] |
+            permission;
+
+        emit PermissionGranted(account, authorized, permission);
+    }
+
+    function _revokePermission(
+        uint256 account,
+        address authorized,
+        uint16 permission
+    ) internal {
+        _accountStore().accountDelegations[account][authorized] =
+            _accountStore().accountDelegations[account][authorized] &
+            ~permission;
+
+        emit PermissionRevoked(account, authorized, permission);
+    }
+
+    function _clearPermissions(uint256 account, address authorized) internal {
+        _accountStore().accountDelegations[account][authorized] = 0;
+
+        emit PermissionsCleared(account, authorized);
+    }
+
+    function _setPermissions(
+        uint256 account,
+        address authorized,
+        uint16[] calldata permissions
+    ) internal {
+        uint16 packedPermissions = 0;
+
+        for (uint i = 0; i < permissions.length; i++) {
+            packedPermissions = packedPermissions | permissions[i];
+        }
+
+        _accountStore().accountDelegations[account][authorized] = packedPermissions;
+
+        emit PermissionsSet(account, authorized, packedPermissions);
+    }
+
+    /// @dev Used to allow certain functions to operate on the account if owner or authorized to that operation
+    modifier onlyOwnerOrAuthorized(uint256 accountId, uint16 permission) {
```

Comment:
This is the modifier that is going to be used on each operation that can be delegated or requires the account owner